amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +154,244 @@ public void stop() {
 
     private void runShardSync() {
         if (leaderDecider.isLeader(workerId)) {
-            for (Map.Entry<StreamIdentifier, StreamConfig> streamConfigEntry : currentStreamConfigMap.entrySet()) {
-                final ShardSyncTaskManager shardSyncTaskManager = shardSyncTaskManagerProvider.apply(streamConfigEntry.getValue());
-                if (!shardSyncTaskManager.syncShardAndLeaseInfo()) {
-                    log.warn("Failed to submit shard sync task for stream {}. This could be due to the previous shard sync task not finished.",
-                             shardSyncTaskManager.shardDetector().streamIdentifier().streamName());
+            log.info(String.format("WorkerId %s is leader, running the periodic shard sync task", workerId));
+            try {
+                // Construct the stream to leases map to be used in the lease sync
+                final Map<StreamIdentifier, List<Lease>> streamToLeasesMap = getStreamToLeasesMap(
+                        currentStreamConfigMap.keySet());
+
+                // For each of the stream, check if shard sync needs to be done based on the leases state.
+                for (Map.Entry<StreamIdentifier, StreamConfig> streamConfigEntry : currentStreamConfigMap.entrySet()) {
+                    final ShardSyncResponse shardSyncResponse = checkForShardSync(streamConfigEntry.getKey(),
+                            streamToLeasesMap.get(streamConfigEntry.getKey()));
+                    if (shardSyncResponse.shouldDoShardSync()) {
+                        log.info("Periodic shard syncer initiating shard sync for {} due to the reason - ",
+                                streamConfigEntry.getKey(), shardSyncResponse.reasonForDecision());
+                        final ShardSyncTaskManager shardSyncTaskManager = shardSyncTaskManagerProvider
+                                .apply(streamConfigEntry.getValue());
+                        if (!shardSyncTaskManager.submitShardSyncTask()) {
+                            log.warn(
+                                    "Failed to submit shard sync task for stream {}. This could be due to the previous pending shard sync task.",
+                                    shardSyncTaskManager.shardDetector().streamIdentifier().streamName());
+                        }
+                    } else {
+                        log.info("Skipping shard sync for {} due to the reason - {}", streamConfigEntry.getKey(),
+                                shardSyncResponse.reasonForDecision());
+                    }
+                }
+            } catch (Exception e) {
+                log.error("Caught exception while running periodic shard syncer.", e);
+            }
```

Comment:
Should we rethrow rather than swallowing? If PeriodicShardSync can't run several times due to some fatal exception and the customer isn't monitoring their logs, they could be stuck with a partial lease table

Reply:
Ya but I believe we should not allow the application to be stopped or interrupted. We can possibly introduce/use metrics.